public/service-worker.js

```diff
@@ -0,0 +1,253 @@
+  pollNotifications();
+});
+
+async function pollNotifications() {
+  // Will not start without a session ID or notification permissions
+  if (!sessionId || Notification.permission !== "granted") {
+    scheduleNextPoll();
+    return;
+  }
+  if (!(await canSendNotification())) {
+    console.log("Notification was recently shown. Skipping notification.");
+    scheduleNextPoll();
+    return;
+  }
+
+  try {
+    const data = await fetchNotificationData();
+    if (data) {
+      // These random numbers will be removed with the new endpoint
+      const randomNumber = Math.floor(Math.random() * 4);
+      const notificationUrl = data[randomNumber].url
+        ? data[randomNumber].url
+        : "/articles";
+
+      self.registration.showNotification("Zeeguu Notification", {
+        body: data[randomNumber].message,
+        icon: "/logo192.png",
+        tag: "zeeguu-notification",
+        data: { url: notificationUrl },
+      });
+      await updateLastNotificationTime();
+    }
+  } catch (error) {
+    console.error("Failed to poll notifications:", error);
+  }
+  scheduleNextPoll();
+}
+
+async function fetchNotificationData() {
+  try {
+    // Temporary json file/endpoint until I've received new endpoint
+    // const response = await fetch(`https://api.zeeguu.org/available_languages`);
+    const response = await fetch(`/mock_notification.json`);
+    if (response.ok) {
+      return await response.json();
+    }
+  } catch (error) {
+    console.error("Failed to fetch notificaiton data", error);
+  }
+}
+
+async function canSendNotification() {
+  try {
+    const cache = await caches.open(NOTIFICATION_CACHE);
+    const cachedNotificationTimestamp = await cache.match(
+      "lastNotificationTime",
+    );
+    const lastNotificationTime = cachedNotificationTimestamp
+      ? await cachedNotificationTimestamp.json()
+      : null;
+
+    const notificationThreshold = 5000;
+
+    const shouldNotify =
+      !lastNotificationTime ||
+      Date.now() - lastNotificationTime >= notificationThreshold;
+
+    return shouldNotify;
+  } catch (error) {
+    console.error("Error checking notification time from cache:", error);
+    return true;
+  }
+}
+
+async function updateLastNotificationTime() {
+  try {
+    const cache = await caches.open(NOTIFICATION_CACHE);
+    const response = new Response(JSON.stringify(Date.now()));
+    await cache.put("lastNotificationTime", response);
+  } catch (error) {
+    console.error("Error updating notification time in cache:", error);
+  }
+}
+
+// Receive notification every 5 seconds (5000 as setTimeout uses milliseconds)
+function scheduleNextPoll() {
+  if (notificationTimer !== null) {
+    clearTimeout(notificationTimer);
+  }
+  notificationTimer = setTimeout(pollNotifications, 5000);
+}
+
+// Used for handling notification clicks
+// https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/notificationclick_event
+self.addEventListener("notificationclick", (event) => {
```

Comment:
In this method you need to check if you can access the ID of the notification.

I am guessing you can since you use the following: `event.notification.data.url` to get the url.